Adding cmdline support submission with helpme

vsoch opened the pull request:
This pull request will address #3649 and #3942 , specifically adding the helpme client (used with GitHub) to support handling the "other errors not known" at the end of the except statements in the datalad/cmdline/main.py.  I developed this interactively and used a different error that I could trigger with:

```bash
datalad --cmd search
```

I wasn't sure what kind of real use case would lead to the last error, so I was only able to get there by commenting out the other exceptions. It would be good to have a few examples to trigger the error so I can test the final thing **and** add an example to the user docs. Interactively, the example issue is here https://github.com/datalad/datalad-helpme/issues/4. 

## Fixes
 - #3649 
 - #3942

## Changes
Changes include:

 - a reporting.rst documentation page to describe helpme and how it works with datalad
 - a submit_helpme function defined in datalad/cmdline/support.py, which doesn't trigger an error if helpme isn't installed (just skips over)
 - adding helpme to install under misc (note that the github and base of helpme only requires requests, which is already required by datalad)
 - an envar `DATALAD_HELPME_DISABLE` that the user can set to disable it.
 - instructions for exporting `HELPME_GITHUB_TOKEN` if they don't want the interactive submission.

Signed-off-by: vsoch <vsochat@stanford.edu

yarikoptic commented on docs/source/reporting.rst:
```diff
@@ -0,0 +1,68 @@
+If you need a completely headless and non-interactive interaction, then you can
+simply export a ``HELPME_GITHUB_TOKEN`` to the environment and it will
+be discovered, and use the GitHub API to open the same issue. See the helpme_
+documentation for more details.
```
Note to ourselves: we could probably integrate with our handling of GitHub token credentials

yarikoptic commented on datalad/cmdline/helpers.py:
```diff
@@ -255,7 +255,6 @@ def get_repo_instance(path=curdir, class_=None):
     else:
         raise RuntimeError("No datalad repository found in %s" % abspath_)
 
-
```
Unrelated change ideally to be reverted (two empty lines too separate at this level I believe)

vsoch replied:
Fixed locally.

yarikoptic commented on datalad/cmdline/support.py:
```diff
@@ -0,0 +1,85 @@
+        # Submit the issue
+        issue = helper.run_headless(repo=repo, body=body, title=title)
+
+    except:
```
Worth catching and logging at debug level this exception

vsoch replied:
Fixed locally.

yarikoptic commented on datalad/cmdline/support.py:
```diff
@@ -0,0 +1,85 @@
+    ]
+
+    try:
+        from helpme.main import get_helper
```
That is where to catch import error,  suggest to install it with an lgr.info message, and then return

vsoch replied:
Fixed locally.

yarikoptic commented on datalad/cmdline/support.py:
```diff
@@ -0,0 +1,85 @@
+
+os.environ["MESSAGELEVEL"] = "QUIET"
+os.putenv("MESSAGELEVEL", "QUIET")
+
```
Why is this process wide upon import change of this environment variables is needed? Ideally should be avoided, or if really needed - a comment why it is needed is needed

vsoch replied:
helpme has a message level controlled via environment, and that's how you disable warnings. if you don't put that there will be an ugly WARNING no GitHub token.... every time the helpme_submit function is used.

vsoch replied:
I'll add a note.

yarikoptic replied:
Then it should be patched only around helpme invocation . I can do that later. I would also recommend in helpme to prepend it with HELPME_ since it sounds to be specific to it

vsoch replied:
Yeah that was bothering me too :) I can do a quick release for that and I'll update the PR here.

vsoch replied:
okay pushing update shortly.

yarikoptic commented on datalad/cmdline/support.py:
```diff
@@ -0,0 +1,85 @@
+    title = title or default_title
+
+    # If the user requests to disable, or in testing environment don't submit
+    if os.environ.get("DATALAD_HELPME_DISABLE") is not None:
```
Note more to ourselves: make it defined in https://github.com/datalad/datalad/blob/master/datalad/interface/common_cfg.py#L373 and use global cfg.getbool here, then it would allow setting it in global or user gitconfig

yarikoptic commented on datalad/cmdline/support.py:
```diff
@@ -0,0 +1,85 @@
+
+def submit_helpme(detail=None, title=None):
+    """Submit a request to the datalad-helpme repository at
+       https://github.com/datalad/datalad-helpme. If helpme isn't installed,
```
Very minor: use that repo variable here?

vsoch replied:
Do you want some called of the function to be able to control the repo? I almost put it there but changed my mind because it seemed more like a global / non-changeable variable.

vsoch replied:
I'll do the same as I did for title.

yarikoptic commented on datalad/cmdline/support.py:
```diff
@@ -0,0 +1,85 @@
+        helper = get_helper(name="github")
+
+        with swallow_outputs() as cmo:
+            wtf(decor="html_details", sections=sections)
```
Note to myself: check if we could just return rendered value avoiding any swallowing

yarikoptic replied:
Checked, can't do ATM but we look into making it possible

yarikoptic commented on docs/source/reporting.rst:
```diff
@@ -0,0 +1,68 @@
+
+To submit to GitHub, HelpMe only requires requests, which is already required by
+Datalad, so the install above will not add any additional dependencies other
+than the helpme library itself. If you don't choose to
```
If there is no additional dependencies if helpme, I would be ok to make it party of the default installation. Let's see what others say though

yarikoptic commented on setup.py:
```diff
@@ -52,6 +52,7 @@
         'PyGithub',          # nice to have
     ],
     'misc': [
+        'helpme[github]',    # support for helpme error submission
```
Let's put current version of helpme as minimally required

vsoch replied:
got it.

vsoch commented:
Quick update - here is the issue submission with traceback, command, and identifier added https://github.com/datalad/datalad-helpme/issues/5. The "this issue is already open" functionality needs to be handled via a GitHub action.

yarikoptic commented on datalad/cmdline/support.py:
```diff
@@ -0,0 +1,105 @@
+       wtf (capturing system and library information) and an optional message.
+
+       Arguments:
+        - title (str)     : the title for the GitHub issue
```
Just spotted it - will need to be adjusted to follow numpy docstring formatting as everywhere else in the code base, I can do that too later

vsoch replied:
I can give this a shot at least.

yarikoptic commented on datalad/cmdline/support.py:
```diff
@@ -0,0 +1,105 @@
+
+    # If no identifier defined, use traceback
+    if identifier is None:
+        identifier = traceback
```
Is there a special meaning of an empty string identifier?

vsoch replied:
We check `if identifier is not None` so it can't be an empty string.

yarikoptic commented on datalad/cmdline/main.py:
```diff
@@ -557,6 +559,12 @@ def main(args=None):
+                submit_helpme(
+                    title = title, 
+                    traceback = traceback.format_exc(),
+                    detail = ' '.join(sys.argv)
```
note 2 @vsoch : pushed minor tune up here (see e.g. deecbc0378588e066609e7442385763c77a0bf15)

vsoch commented:
@yarikoptic I made a helpme release that honors the environment namespace with `HELPME_MESSAGELEVEL` and moved the setting down into the function, also made an attempt at the docstring. https://github.com/datalad/datalad/pull/4586/commits/11551e783649bac7ecc08aafeeca1daf58576cc5

vsoch commented:
okay I'm peacing out for today - let me know next steps! I think it would be good to know some commands that I could do to trigger this "not understood error" so we can test helpme. I think we need to do:

 - test with GitHub token
 - test with GitHub token via GitHub action (so it can run with the CI)
 - test to open issue on datalad-helpme
 - test GitHub workflow for datalad-helpme https://github.com/datalad/datalad-helpme/pull/7

Peeeace! and Peas.

yarikoptic commented on datalad/cmdline/main.py:
```diff
@@ -557,6 +560,12 @@ def main(args=None):
                 sys.exit(exc.code if exc.code is not None else 1)
             except Exception as exc:
+                title = "DataLad Cmd Error: %s" % exc.__class__.__name__
+                submit_helpme(
```
Myself: need to check if actually current UI  is interactive (UI while running tests is not, so that would avoid setting envvar in Cis). We might actually make env var such that it could force helpme in noon interactive sessions to assist filling issues in such cases

yarikoptic replied:
And ideally in the longer run would be great if helpme could use our ui interface to ask questions etc

vsoch replied:
Definitely! We can handle this in another scoped PR / issue.

yarikoptic commented on datalad/cmdline/main.py:
```diff
@@ -557,6 +560,12 @@ def main(args=None):
                 # had no code defined
                 sys.exit(exc.code if exc.code is not None else 1)
             except Exception as exc:
+                title = "DataLad Cmd Error: %s" % exc.__class__.__name__
```
I think we can strip this coming prefix (no unique info), add a new label eg helpme or autofiled, and include instead limited in length str(exc) .

vsoch replied:
Fixed locally.

vsoch commented:
@yarikoptic I've updated helpme to allow using a custom generated identifier, and figured out how to extract a list of (non datalad) and datalad functions, and updated here to generate hashes for each, along with the exception name. See the dummy issue at https://github.com/datalad/datalad-helpme/issues/9. Once we have decided that this identifier / format looks okay I'll update the github workflow PR on the datalad-helpme respository to handle parsing it.

mih commented:
Closing this after 6+ months of inactivity. I have placed an agenda item for next week's developer call to discuss a future of this feature. It certainly is not desirable to have it sit in an open PR for months.
